roles/openshift_cloud_provider/tasks/main.yml

```diff
@@ -1,3 +1,14 @@
 ---
+- name: Set cloud provider facts
+  openshift_facts:
+    role: cloudprovider
+    openshift_env: "{{ hostvars[inventory_hostname] | oo_openshift_env }}"
+    openshift_env_structures:
+    - 'openshift.cloudprovider.aws.*'
+    - 'openshift.cloudprovider.openstack.*'
+
 - include: openstack.yml
-  when: "openshift_cloud_provider is defined and openshift_cloud_provider == 'openstack' and 'provider' in openshift and openshift.provider.name == 'openstack'"
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind == 'openstack' and 'openstack' in openshift.cloudprovider"
+
+- include: aws.yml
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind == 'aws'"
```

Comment:
Although, my defaults for cloudprovider should leave us with at least a kind of None.